benchmarks/common.sh

```diff
@@ -1,83 +1,193 @@
+  echo "    LLVM_DIR=/path/to/llvm/build $*"
   echo "Otherwise, just install the clang package"
-  exit
+  exit 1
 fi
 
-# Clang.
-echo "Using Clang in: "
-which clang
+LOG "Running parameters:"
+LOG "    Compiler: $CLANG"
+HARNESS="$ROOT_DIR/benchmarks/harness/controller.py"
+LOG "     Harness: ${HARNESS}"
+TEST_PATH="$ROOT_DIR/test/Benchmarks"
+LOG "   TEST path: ${TEST_PATH}"
 
-# Assembler.
-echo "Using Assembler in: "
-which llc
+# Now we set all basic paths
+LIB_PATH="$BUILD_DIR"/lib
+if [ "$LIBXSMMROOT" ]; then
+  LIB_INCLUDE_PATH="$LIBXSMMROOT"/include
+else
+  LIB_INCLUDE_PATH="$BUILD_DIR"/_deps/xsmm-src/include
+fi
+unamestr=$(uname)
+if [[ "$unamestr" == 'Darwin' ]]; then
+  export DYLD_LIBRARY_PATH=$LIB_PATH
+else
+  export LD_LIBRARY_PATH=$LIB_PATH
+fi
+
+LOG "    LIB path: ${LIB_PATH}"
+LOG "LIB INC path: ${LIB_INCLUDE_PATH}"
+LOG ""
 
-# LLVM MLIR IR to LLVM IR.
-echo "Using mlir translate in: "
-which mlir-translate
+RED='\033[0;31m'
+GREEN='\033[0;32m'
+NC='\033[0m' # No Color
 
-# TPP compiler.
-echo "Using tpp compiler in: "
-which tpp-opt
+# Common functions
+
+# run_reference(source, num_it, *)
+#
+# - source: C source file, mandatory
+# - num_it: number of iteratios, mandatory
+# - *: extra arguments to the tool
+#
+# Note: A common extra arg is -x (for xsmm runs)
+function run_reference() {
+  SRC="$1.c"
+  if [ ! -f "$SRC" ]; then
+    >&2 echo "ERROR: File '$SRC' not found"
+    exit 1
+  fi
+  DST="${SRC%.c}"
+  ITER=$(($2+0))
+  if [ "$ITER" == "0" ]; then
+    >&2 echo "ERROR: Invalid iteration '$2'"
+    exit 1
+  fi
+  shift 2
+  ARGS="$*"
+  CMD="$CLANG -O3 $SRC -L\"$LIB_INCLUDE_PATH\" -o $DST"
+  if $CMD &> /dev/null; then
+    # If it compiles, run
+    CMD="./$DST -i $ITER $ARGS"
+    if ! $CMD; then
+      >&2 echo "ERROR: Running $DST"
+      >&2 echo "   $CMD"
+      exit 1
+    fi
+  else
+    # Else, just bails
+    >&2 echo "ERROR: Cannot compile '$SRC'"
+    >&2 echo "   $CMD"
+    exit 1
+  fi
+}
+
+# run_tpp_mlir(source, num_it, alt, *)
+#
+# - source: MLIR source file, mandatory
+# - num_it: number of iteratios, mandatory
+# - alt: alternative name, if any (complement source), optional
+# - *: extra arguments to the tool
+#
+# Note: alt name is "$source_$alt.mlir" (ex. matmul_12x24x48.mlir")
+function run_tpp_mlir() {
+  SRC="$TEST_PATH/$1.mlir"
+  ITER=$(($2+0))
+  if [ "$ITER" == "0" ]; then
+    >&2 echo "ERROR: Invalid iteration '$2'"
+    exit 1
+  fi
+  ALT_SRC="$TEST_PATH/$1_$3.mlir"
+  if [ ! -f "$SRC" ]; then
+    # Try with extension
+    if [ -f "$ALT_SRC" ]; then
+      SRC=$ALT_SRC
+      shift
+    else
+      >&2 echo "ERROR: File '$SRC' not found"
+      if [ -n "$3" ]; then
+        >&2 echo "ERROR: File '$ALT_SRC' also not found"
+      fi
+      exit 1
+    fi
+  fi
+  shift 2
+  ARGS="$*"
+  CMD="$HARNESS -q $SRC -n $ITER $ARGS"
+  if ! $CMD; then
+    >&2 echo "ERROR: Harness failed"
+    >&2 echo "   $CMD"
+    exit 1
+  fi
+}
+
+# run_benchmark()
```

Comment:
nit: mix of positional and getopts arguments makes UX a bit confusing

Reply:
It is, and I intend to fix it, but it's not easy when the arguments that I want to pass down are different. Note that there is no UX here, this is mostly to drive the benchmarks via CI, and help somewhat reproduce the errors.

Users should really be using the harness (for the MLIR part) and the C part should be stable and consistent, so no need to modify or inspect the results.

Reply:
Fair point. We might want to tweak the baseline in the future but there’s no substantial gain in making it particularly pretty at the backend.